### ISSUE TITLE
Using tf.Dataset in non-eager mode impossible

### ISSUE THREAD

**Issue by Flamefire** [tensorflow/tensorflow#38748]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): yes
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): 2.1.0
- Python version: 3.7.4

**Describe the current behavior**

I'd like to compare eager and non-eager mode using a custom training loop and a TF Dataset as the input.

This seems to be impossible as shape information seems to be unavailable and even basic training loops throw "Attempting to capture an EagerTensor without building a function."

I also couldn't find any documentation on how to do that.

**Describe the expected behavior**

Custom training loops should be possible in graph (non-eager) mode with tf.Dataset and documented

**Standalone code to reproduce the issue**
Heavily reduced example:

```
import tensorflow as tf
tf.compat.v1.disable_eager_execution()

def preprocess(x, l):
    return tf.image.convert_image_dtype(x, tf.float32), l

train_data, test_data = tf.keras.datasets.mnist.load_data()
train_data = tf.data.Dataset.from_tensor_slices(train_data).map(preprocess)

@tf.function
def run_loop(model, data):
    res = True
    for x, y in data:
        batch_size = int(x.shape[0])
        res = model(x)[0] == 1.
    return res

model = tf.keras.Sequential([tf.keras.layers.Dense(10, activation='softmax')])
result = run_loop(model, train_data.batch(32, drop_remainder=False)) # drop_remainder=True for "EagerTensor" exception
```

**Other info / logs** 

The first failure in the above is the line with `batch_size` as it tries to convert a `None` value, but at that point the value should be known already.
Changing the default `drop_remainder` to `True` avoids this.

The tensor causing the "capture..." failure has a dtype of `resource`. Not sure what that is.

Obviously the above doesn't do much, but reproduces the failure. My real training loop has all the GradientTape, loss, callbacks etc.

**Comment by Saduf2019** [comment]
@Flamefire 
i ran the code shared by you and face this error,please find the [gist here](https://colab.sandbox.google.com/gist/Saduf2019/dcf9229cae9d4ee245ee7907ed73c4d6/untitled145.ipynb)

**Comment by Flamefire** [comment]
@tensorflow-butler not solved

**Comment by Saduf2019** [comment]
@Flamefire
Please update as per above comment

**Comment by jsimsa** [comment]
@tomerk could you please take a look (or triage to someone on the Keras team)? Thank you.

**Comment by tomerk** [comment]
Hi @Flamefire,

A few things here:
1. Why are you trying to use a custom training loop *and* disabling eager execution?  Generally speaking if you're using a custom training loop (as opposed to using tf1 session-based code) then disabling eager execution probably isn't the right way to achieve whatever you're trying to achieve. If you're trying to get graph performance using tf.function will do that. But, I see you're using a tf.function *and* calling compat.v1 to disable eager execution. So what's the goal here?

Keep in much of TF2 won't work properly if you've explicitly disabled eager. This can pop up with pretty unexpected error messages

2. The tensor's `.shape` attribute actually  *isn't*  known at that point in time.
`tensor.shape` is a static property on the tensor that is set at tensor construction time and can always be accessed as a direct python object. When you don't drop the remainder the batch size is unknown and will be `None` in the static property.

If you want to dynamically grab the batch size and use it in your computation you need to use `tf.shape(x[0])`. `tf.shape` returns a *tensor* representing whatever the shape of the tensor is at the point when it's used.

From the documentation of `tensor.shape`:
```
    """Returns a `tf.TensorShape` that represents the shape of this tensor.

    >>> t = tf.constant([1,2,3,4,5])
    >>> t.shape
    TensorShape([5])

    `tf.Tensor.shape` is equivalent to `tf.Tensor.get_shape()`.

    In a `tf.function` or when building a model using
    `tf.keras.Input`, they return the build-time shape of the
    tensor, which may be partially unknown.

    A `tf.TensorShape` is not a tensor. Use `tf.shape(t)` to get a tensor
    containing the shape, calculated at runtime.

    See `tf.Tensor.get_shape()`, and `tf.TensorShape` for details and examples.
    """
```

3. I was originally seeing the "capture EagerTensor" exception you mentioned, but it no longer seems to be appearing for me... do you see it in the tf-nightly & with the batch size access fixed?

Keep in mind it's very low priority for us to fix if you're using `compat.v1.disable_eager_execution` unless it's an active regression from TF1 running the same code.

**Comment by Flamefire** [comment]
@tomerk Thanks for your response

1. As written in the OP I wanted to compare eager and non-eager execution of the same code. From your response I conclude that using the v2 APIs this is simply not intended and hence impossible and only worked by chance. Maybe this could be made more clear in the documentation? From the current description I understood that "eager" is the default which trades off performance for ease of use. I hence wanted to see what happens when it is turned off and expected a performance improvement. The problem with tf.function is, that it does not fully recover graph performance. Likely due to synchronization happening after each batch to get the metric values from the GPUs and maybe due to different data handling (preprocessing, uploading, ...). Also some callbacks can't be run inside a tf.function so the whole training loop can't be wrapped in a tf.function but only the individual steps

2. Thanks for the clarification. It seems I used that previously outside of a tf.function where it worked. The quoted part of the documentation is pretty clear. Haven't seen that previously or it was updated since.

3. Can confirm this is fixed in the nightly.

**Comment by tomerk** [comment]
The way to think about TF2 is it's eager (by default), and you can enter a `tf.function` to run part of your code in a graph for performance reasons.

`tf.compat.v1.disable_eager_execution` isn't designed for entering a graph for performance reasons, it's meant as a totally-legacy-compatibility toggle if you find yourself in a situation where you need to run graph-only TF1 code with TF2 binaries. This might happen e.g. if you're actively migrating code from TF1 to TF2, or running legacy code in an environment that only has TF2 installed.

From the TF api docs for compat.v1.disable_eager_execution:
```
This function can only be called before any Graphs, Ops, or Tensors have been created. It can be used at the beginning of the program for complex migration projects from TensorFlow 1.x to 2.x.
```
Please let us know how we could clarify this, or if it was a different piece of documentation that led you to believe that to build graphs in TF2 you are supposed to call `disable_eager_execution()`
--------------

Generally speaking if code doesn't run in a tf.function it would not run in a legacy TF 1.x graph at all, so that's unlikely to be the cause of tf.function not matching TF 1.x graph performance. E.g. datasets can be constructed (or passed into) tf.functions, so it's possible to iterate over your entire dataset in a tf.function call.

Callbacks don't usually have to be called every training step, so you can roll several training steps into one tf.function and only run callbacks in between them. If you're writing a custom loop it's possible to do yourself in your loop. As of recently, Keras models have a `_make_train_function` you can override to do just when you're fitting the model with `model.fit()`. (Guides for this are in progress I believe, or @omalleyt12 should be able to point you to some). These callbacks wouldn't run inside a TF 1.x graph regardless.

That's not to say `tf.function` never has overheads. Depending on how you write your code, you might accidentally cause the function to retrace/recompile more than you expected. If that's the case you should look at the "better performance" w/ tf.functions guide:
https://www.tensorflow.org/guide/function
Any part of your code that is inside a single tf.function trace should run just as fast (or faster) w/ TF2 as it would in a TF1 graph though. (and for things that don't feel free to file a bug!)

**Comment by tomerk** [comment]
Closing this issue for now because the reported error is fixed in nightly.
